Mod1/Lessons/Week5/Scope.md

```diff
@@ -0,0 +1,232 @@
+# Scope and Access Modifiers
+
+## Learning Goals
+* Identify how **scope** affects where variables can be accessed
+* Have a reference to look at when **scope** errors show up
+* Understand the difference between the **public** and **private** access modifiers
+
+# Scope
+
+## Intro
+Scope is what you have access to and where you have access to it.
+
+Scope is the kind of thing that you think you don’t need to know until it starts causing you trouble. Even then, it is generally unnecessary to know the full rules. It’s more important that when you see unexpected behavior scope of your variables/method names is one of the things on your list to check.
+
+## Examples of Scope Errors
+
+You've probably already run into scope errors.
+
+Maybe you've seen something like this where you already defined a variable but are still getting an error saying your variable doesn't exist.
+
+Example 1:
+![Scope Example 1](/Mod1/Images/Week5/ScopeExample1.png)
+
+Or maybe you've tried to use the variable i multiple times in your program and found that you need to create a new variable each time.
+
+Example 2:
+![Scope Example 2](/Mod1/Images/Week5/ScopeExample2.png)
+
+These errors are both caused by scope!
+
+## Scope Definition
+
+I like this definition from a [Microsoft Tutorial](https://learn.microsoft.com/en-us/training/modules/csharp-code-blocks/2-exercise-variable-scope).
+
+> Variable scope is the **visibility of the variable** to the other code in your application. A locally scoped variable is only accessible inside of the code block in which it's defined. If you attempt to access the variable outside of the code block, you'll get a compiler error.
+
+<!-- Should I try and define a code block here? -->
+
+With this new definition in mind, we can see why the above code examples caused errors.
+
+Example 1:
+![Scope Example 1 With Highlighting](/Mod1/Images/Week5/ScopeExample1Highlighting.png)
+
+We are trying to access the `username` variable outside of the scope where it is defined.
+
+Example 2:
+![Scope Example 2 With Highlighting](/Mod1/Images/Week5/ScopeExample2Highlighting.png)
+
+Note that a variable is accessible **everywhere** inside the code block in which it's defined, even from within a nested code block.
+
+## Practice
+
+Here is a modified version of some code from our Class Interaction Lesson. 
+
+>Take 3 minutes: For each variable defined, determine the box for the scope where that variable is available.
+
+<details><summary>Solution</summary><br/>
+
+![Scope Example 3 With Highlighting](/Mod1/Images/Week5/ScopeExample3.png)
```

Comment:
I realized while working on this that it might not be readable to folks that are colorblind. It's already a lot going on for anyone to look at. Let me know if you have suggestions for readability here!

Reply:
Love that you're thinking about this!

One thing that might help is to use different outlining for each box - not sure how easy that might be, but we could do something like outer box dashed, next box dotted, next box solid, next box dashed. just another visual differentiator between each box.

Reply:
I'll give that a go! It's easy in excalidraw to use different borders. I'll take a look to see how visable the different borders are.